modules/crud/sql/update.go

```diff
@@ -1,60 +1,226 @@
+	}
+	switch req.Operation {
+	case utils.All:
+		for k := range req.Update {
+			switch k {
+			case "$set", "$inc", "$mul", "$max", "$min", "$currentDate":
+				sqlQuery, args, err := s.generateUpdateQuery(ctx, project, col, req, k)
+				if err != nil {
+					return err
+				}
+				err = doExecContext(ctx, sqlQuery, args, executor)
+				if err != nil {
+					return err
+				}
+			default: // (case "$push", "$unset", "$rename")
+				return utils.ErrInvalidParams
+			}
+		}
+	case utils.Upsert:
+		for k := range req.Update {
+			switch k {
+			case "$set", "$inc", "$mul", "$max", "$min", "$currentDate":
+				err := s.upsert(ctx, project, col, req, k)
+				if err != nil {
+					return err
+				}
+			default: // (case "$push", "$unset", "$rename")
+				return utils.ErrInvalidParams
+			}
+		}
+	default: // (case utils.One)
+		return utils.ErrInvalidParams
+	}
+	return nil
 }
 
-//generateUpdateQuery makes query for update operation
-func (s *SQL) generateUpdateQuery(ctx context.Context, project, col string, req *model.UpdateRequest) (string, []interface{}, error) {
+//generateUpdateQuery makes query for update operations
+func (s *SQL) generateUpdateQuery(ctx context.Context, project, col string, req *model.UpdateRequest, op string) (string, []interface{}, error) {
 	// Generate a prepared query builder
-	query := goqu.From(col).Prepared(true)
-	query = query.SetAdapter(goqu.NewAdapter(s.dbType, query))
+	dialect := goqu.Dialect(s.dbType)
+	query := dialect.From(col)
+	if op == "$set" {
+		query = query.Prepared(true)
+	}
 
 	if req.Find != nil {
-
 		// Get the where clause from query object
 		var err error
 		query, err = generateWhereClause(query, req.Find)
 		if err != nil {
 			return "", nil, err
 		}
-
 	}
 
 	if req.Update == nil {
 		return "", nil, utils.ErrInvalidParams
 	}
+	m, ok := req.Update[op].(map[string]interface{})
+	if !ok {
+		return "", nil, utils.ErrInvalidParams
+	}
 
-	record, err := generateRecord(req.Update["$set"])
+	if op == "$currentDate" {
+		for k,v := range m {
+			mm, ok := v.(map[string]interface{})
+			if !ok {
+				return "", nil, utils.ErrInvalidParams
+			}
+			for _, val := range mm {
+				val, ok := val.(string)
+				if !ok {
+					return "", nil, utils.ErrInvalidParams
+				}
+				switch val {
+				case "date", "timestamp":
+					m[k] = val
+				default:
+					return "", nil, utils.ErrInvalidParams
+				}
+			}
+		}
+	}
+	
+	record, err := generateRecord(req.Update[op])
 	if err != nil {
 		return "", nil, err
 	}
 
 	// Generate SQL string and arguments
-	sqlString, args, err := query.ToUpdateSql(record)
+	sqlString, args, err := query.Update().Set(record).ToSQL()
 	if err != nil {
 		return "", nil, err
 	}
 
 	sqlString = strings.Replace(sqlString, "\"", "", -1)
+	switch op {
+	case "$set":
+	case "$inc":
+		for k, v := range m {
+			val, err := numToString(v)
+			if err != nil {
+				return "", nil, err
+			}
+			sqlString = strings.Replace(sqlString, k+"="+val, k+"="+k+"+"+val, -1)
```

Comment:
This is prone to injection attacks.
Why aren't we using a transaction? Such operations aren't truly atomic in all cases. We'll explicitly set the consistency level to make sure we have a record level lock so no other query can simultaneously perform a read or write.